File: codeforces-practice/1600s/maze.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <set>
#include <queue>

using namespace std;

vector<pair<int,int>> get_neighbours(int y, int x, int height, int width)
{
    vector<pair<int,int>> output;
    if (y >= 1)
        output.push_back({y-1, x});
    if (x >= 1)
        output.push_back({y, x - 1});
    if (x < width - 1)
        output.push_back({y, x  + 1});
    if (y < height - 1)
        output.push_back({y+1, x});
    return output;
}

int main()
{
    unsigned int n, m, k;
    cin >> n >> m >> k;

    vector<string> matrix(n);

    pair<int,int> starter;
    unsigned int period_count = 0;
    for (int y = 0; y < n; y++)
    {
        string input;
        cin >> input;
        matrix[y] = input;

        for (int x = 0; x < m; x++)
        {
            if (matrix[y][x] == '.')
            {
                period_count += 1;
                starter = {y,x};
            }
        }
    }
    queue<pair<int,int>> period_queue({starter});

    unsigned int keeping = period_count - k;
    unsigned int kept = 0;
    set<pair<int,int>> visited;

    while (kept < keeping)
    {
        auto coords = period_queue.front();
        period_queue.pop();

        if (visited.count(coords)) continue;

        kept += 1;
        visited.insert(coords);

        for (auto [new_y, new_x] : get_neighbours(coords.first, coords.second, n, m))
        {
            if (matrix[new_y][new_x] == '.')
            {
                period_queue.emplace(new_y, new_x);
            }
        }
    }


    for (int i = 0; i < n; i++)
    {
        for (int j = 0; j < m; j++)
        {
            if (visited.count({i, j}) == 0 && matrix[i][j] == '.')
            {
                cout << "X";
            }
            else
            {
                cout << matrix[i][j];
            }
        }
        cout << "\n";
    }
}

File: codeforces-practice/1600s/maze.py
from collections import defaultdict, deque

n, m, k = map(int, input().split())

matrix = []
for _ in range(n):
    matrix.append(input())


def get_neighbours(y: int, x: int, height: int, width: int) -> list[(int,int)]:
    output = []
    if y >= 1:
        output.append((y-1, x))
    if y < height-1:
        output.append((y+1, x))
    if x >= 1:
        output.append((y, x-1))
    if x < width-1:
        output.append((y, x+1))
    return output

def solve(matrix, k):
    height = len(matrix)
    width = len(matrix[0])

    count_of_periods = 0 
    starter = None
    for y in range(height):
        for x in range(width):
            c = matrix[y][x]
            if c == '.':
                count_of_periods += 1
                starter = (y,x)
    keep_amount = count_of_periods - k 

    stack = [starter]

    kept = 0
    visited = set()
    while kept < keep_amount:
        item = stack.pop()
        y = item[0]
        x = item[1]
        if (item in visited or matrix[y][x] != '.'): continue

        visited.add((item))
        kept += 1
        for n_y, n_x in get_neighbours(y, x, height, width):
            stack.append((n_y, n_x))


    for y in range(height):
        for x in range(width):
            if (matrix[y][x] == '.' and (y,x) not in visited):
                print('X', end='')
                continue
            print(matrix[y][x], end='')
        print()

solve(matrix, k)
    